src/main/java/org/intellij/sdk/codesync/clients/CodeSyncClient.java

```diff
@@ -0,0 +1,146 @@
+    }
+
+    public Integer uploadFile(ConfigRepo configRepo, DiffFile diffFile, File originalsFile) throws FileInfoError, InvalidJsonError, RequestError {
+        HttpClient httpClient = HttpClientBuilder.create().build();
+        HttpPost post = new HttpPost(this.filesURL);
+        JSONObject data = new JSONObject();
+        Map<String, Object> fileInfo;
+        try {
+            fileInfo = Utils.getFileInfo(originalsFile.getAbsolutePath());
+        } catch (FileInfoError error) {
+            throw new FileInfoError(String.format("File Info could not be found for %s", diffFile.fileRelativePath));
+        }
+
+        data.put("repo_id", configRepo.id);
+        data.put("branch", diffFile.branch);
+        data.put("is_binary", (Boolean) fileInfo.get("isBinary"));
+        data.put("size", (long) fileInfo.get("size"));
+        data.put("file_path", diffFile.fileRelativePath);
+        data.put("created_at", Utils.formatDate(diffFile.createdAt));
+
+        StringEntity dataStr;
+        try {
+            dataStr = new StringEntity(data.toJSONString());
+        } catch (UnsupportedEncodingException error) {
+            throw new InvalidJsonError("Invalid JSON encoding error raised while uploading file to the server.");
+        }
+
+        post.setEntity(dataStr);
+        post.setHeader("Content-type", "application/json");
+        post.setHeader("Authorization", String.format("Basic %s", configRepo.token));
+
+        HttpResponse response;
+        try {
+            response = httpClient.execute(post);
+        } catch (IOException e) {
+            throw new RequestError(String.format("Error uploading file. Error: %s", e.getMessage()));
+        }
+
+        String responseContent;
+        try {
+            responseContent = EntityUtils.toString(response.getEntity());
+        } catch (IOException | org.apache.http.ParseException error) {
+            throw new RequestError(String.format("Error processing response of the file upload  request. Error: %s", error.getMessage()));
+        }
+
+        Long fileId;
+        JSONObject jsonResponse;
+        try {
+            jsonResponse = (JSONObject) JSONValue.parseWithException(responseContent);
+            fileId = (Long) jsonResponse.get("id");
+        } catch (org.json.simple.parser.ParseException | ClassCastException error) {
+            throw new RequestError(String.format("Error processing response of the file upload  request. Error: %s", error.getMessage()));
+        }
+
+        Map<String, Object> preSignedURLData;
+        try {
+            preSignedURLData = (Map<String, Object>) jsonResponse.get("url");
+
+            long fileSize =  (long) fileInfo.get("size");
+            if (fileSize > 0) {
+                this.uploadToS3(originalsFile, preSignedURLData);
+            }
+        } catch (ClassCastException error) {
+            // this would probably mean that `url` is empty and we can skip aws upload.
```

Comment:
Maybe can add `url is empty` check in above so that only specific expection is raised

Reply:
Since, "url" will either be a string or an object containing some value. I think, it is not possible to avoid `ClassCastException`.

Reply:
cool 👍